### PR TITLE
Refactor auth + add OIDC

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
Authentication

### Description
This PR refactors auth:

- Removal of all auth provider stubs except Local username/password, LDAP and OIDC
- Add support for OIDC and OAuth2 auth providers + redirect on logout
- Add account creation page for first login with external auth provider
- Check username on register page and account creation page for conflicts while typing
- Increase minimum password length from 1 to 6 characters
- Limit usernames to min 3 and max 64 characters
- Change config option for session lifetime from milliseconds to seconds
- Add config option to forbid profile data changing
- Update docs
- Add config option to set sync source
- Update profile data on sign-in via sync-source

### Screenshot

image

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)
part of #1006

### CODE REVIEW

**Comment by n-connect** [comment]
@ErikMichelson 

Great news! How will this PR will change the OICD client capabilities in general? I've read #1152 , so OIDC Discovery will added too.
For example since [Authelia v4.38.x](https://www.authelia.com/blog/4.38-release-notes/#client-authentication-method-token-endpoint), the following guide needs an update [OAuth with Authelia SSO (self-hosted)](https://docs.hedgedoc.org/guides/auth/authelia/), to add:
`        token_endpoint_auth_method: client_secret_post` 
before `audience: []` , or somewhere near policy.

Do you plan to add additional authentication methods beside "client_secret_post" like "client_secret_basic" or the JWT ones, as listed in [OpenID Client Sends Code](https://openid.net/specs/openid-connect-basic-1_0.html#TokenRequest) or in Authelia [OpenID Connect 1.0 clients | Configuration | Authelia](https://www.authelia.com/configuration/identity-providers/openid-connect/clients/#token_endpoint_auth_method)?

**Comment by ErikMichelson** [comment]
Hi @n-connect, thanks for your thoughts.

We're still in the process of experimenting and refactoring with some auth methods, so I can't answer your questions definitely for now.
OIDC Discovery will be added very probably, regarding the Authelia config and additional methods I can't say anything yet, sorry.

File: frontend/src/api/users/index.ts [hunk 1]
```diff
@@ -14,6 +14,6 @@ import type { UserInfo } from './types'
  * @throws {Error} when the api request wasn't successful.
  */
 export const getUser = async (username: string): Promise<UserInfo> => {
```
Comment (InnayTool) [inline frontend/src/api/users/index.ts:16]:
```suggestion
export const getUserInfo = async (username: string): Promise<UserInfo> => {
```

File: frontend/src/api/auth/index.ts [hunk 2]
```diff
@@ -4,12 +4,29 @@
...
+  return response.asParsedJsonObject()
+}
+
+/**
+ * Requests to check if a username is available.
+ *
+ * @param username The username to check.
+ * @returns {boolean} whether the username is available or not.
+ */
+export const checkUsernameAvailable = async (username: string): Promise<boolean> => {
```
Comment (InnayTool) [inline frontend/src/api/auth/index.ts:26]:
```suggestion
export const checkUsernameAvailablability = async (username: string): Promise<boolean> => {
```

File: frontend/src/api/auth/index.ts [hunk 3]
```diff
@@ -4,12 +4,29 @@
...
+import type { LogoutResponseDto, UsernameCheckDto, UsernameCheckResponseDto } from './types'
 
 /**
  * Requests to log out the current user.
  *
  * @throws {Error} if logout is not possible.
  */
-export const doLogout = async (): Promise<void> => {
-  await new DeleteApiRequestBuilder('auth/logout').sendRequest()
+export const doLogout = async (): Promise<LogoutResponseDto> => {
```
[the remark this answers is not in the document]
Comment (ErikMichelson) [reply frontend/src/api/auth/index.ts:15]: Same as with login, I'd like to clarify the action-aspect of this method which is more clear with the "do".

**Review by ErikMichelson — COMMENTED** [review]
I created issues for the open TODOs. From my side this PR looks ready, I just self-reviewed the code again and didn't find any problematic things anymore.

**Pull request merged**